### PR TITLE
"Edit a field" feature in schema editor

### PR DESCRIPTION
**Pull request by jurgenwerk**
This feature allows editing a field. There is an option to do so from the contextual menu in the schema editor. 

https://github.com/cardstack/boxel/assets/273660/30d1aece-bb34-4f23-9c69-d24ad842f966

### CODE REVIEW

**Comment by tintinthong** [comment]
Hmmm does saving after edit take a long time?? I seem to need to wait for quite a while

EDIT: Owh is this actually the dependencies breaking during edit??

**Comment by tintinthong** [comment]
If we treat edit as a remove and add, then the order of the fields seem to change. At first sight, I think order should be maintained

**Comment by tintinthong** [comment]
https://github.com/cardstack/boxel/assets/8165111/735ec748-9007-4aeb-8335-b2a5d6ed6eda

Exported name seems problematic when trying to edit a parent card

**Comment by jurgenwerk** [comment]
> Hmmm does saving after edit take a long time?? I seem to need to wait for quite a while

How long? It is around 1 second for me.

> Owh is this actually the dependencies breaking during edit??

What do you mean?

> If we treat edit as a remove and add, then the order of the fields seem to change. At first sight, I think order should be maintained

Agree, will look into this

> Exported name seems problematic when trying to edit a parent card

Good catch. Hmm, I think this is because the `moduleSyntax` is actually from the active loaded file in the url bar (the topmost item in the schema editor). If we want to have these options for the parent card (add field, remove field, edit field) then I think we'll have to load the files for every card in the inheritance chain

**Comment by jurgenwerk** [comment]
Discussed with @lukemelia the possibility of allowing field operations only for the schema from the currently opened file (the topmost item in the schema editor). This would simplify things by not having to load the module of the parent card, but maybe also increase user confusion to see the field operations are not available for parent items in schema editor (cards that are lower - under the "inherited from" markets). Will try to make a design proposal to address and get more opinions

**Comment by jurgenwerk** [comment]
Regarding the bug where editing fields from cards outside the currently opened module - I opted into disabling add/remove/edit for such fields. 

Regarding the messed up order after saving - I have a working solution, just needs a little cleanup and a test. Will push a commit for this soon.

**Comment by tintinthong** [comment]
https://github.com/cardstack/boxel/assets/8165111/29f72710-1ab1-4da0-b5b5-b8849aab5436

Possible index error. The 1st field seems to be yield error on edit. Im using staging preview to test

**Comment by jurgenwerk** [comment]
Hold off with the review, looks like something went really wrong after merging with upstream and now the modal is rendered inside the column. Need to inspect

**Comment by jurgenwerk** [comment]
Okay new issues discovered here. 

Regarding the out of place modal, I think this PR might have broke something https://github.com/cardstack/boxel/pull/754? (cc: @backspace) - i tested this by reverting the commit locally and it started working again so I suspect something broke there.

Regarding the issued with editing the first field, there is a bug - need to take care of this special case. Working on a fix.

**Comment by tintinthong** [comment]
> Okay new issues discovered here.
> 
> Regarding the out of place modal, I think this PR might have broke something #754? (cc: @backspace) - i tested this by reverting the commit locally and it started working again so I suspect something broke there.

Is this the matter of needing to rebuild `boxel-ui/addon`??

**Comment by jurgenwerk** [comment]
I do not think so. This issue is present in staging too:

image

**Comment by jurgenwerk** [comment]
After merging main I discovered another problem - after editing a field there is now a bunch of errors around the field not being found in the areas of indexing, card loading, and serializing. Looks like there were some changes upstream that started causing that. We have a ticket to not throw errors when there are fields in a card instance that are different from what is found in the definition. I'll start working on that

**Comment by backspace** [comment]
The narrowed modal goes away if I remove [this line](https://github.com/cardstack/boxel/blob/b15f78ad27da8a43e752061c9da242a0714233a1/packages/boxel-ui/addon/src/components/resizable-panel-group/panel.gts#L72) which is what the container query needs to determine which container length to watch. I’ll open an issue for this and start on it now.

**Comment by jurgenwerk** [comment]
This PR is now considered blocked until "Do not error when there are extra fields in a card instance" ticket is done. Because of that, saving does not work correctly because the server will respond with indexing errors. Will probably fix that in another PR then come back to this one.

**Comment by tintinthong** [comment]
> This PR is now considered blocked until "Do not error when there are extra fields in a card instance" ticket is done. Because of that, saving does not work correctly because the server will respond with indexing errors. Will probably fix that in another PR then come back to this one.

**Comment by jurgenwerk** [comment]
The ordering issue is now fixed. The fields stay in their position after edit. 

Also reindexing/loading of the card doesn't break anymore when the definition changes and we end up with a broken instance. This because we're missing a data migration step and/or versioning, but this is a separate task that requires more planning. For now we make the deserializer more tolerant to field mismatch so that the client can still show the broken instance data and give the user a chance to migrate fields manually (after editing them in the schema editor). 

Ready for another review.

File: packages/base/card-api.gts
Comment (tintinthong) [inline packages/base/card-api.gts:2336]: 👍

**Review by tintinthong — APPROVED** [review]
Tested works as I expected

File: packages/host/app/components/operator-mode/edit-field-modal.gts
Comment (lukemelia) [inline packages/host/app/components/operator-mode/edit-field-modal.gts]:
```suggestion
  private setInitialValuesForExistingField = restartableTask(async () => {
```

**Review by lukemelia — APPROVED** [review]

**Review by burieberry — APPROVED** [review]
Looks good. One suggestion is leaving some extra space under the input so the error message appeared there without pushing the field below it up and down.

**Review by habdelra — APPROVED** [review]

**Pull request merged**
